### Why use Variance and Standard Deviation?

- They show how **spread** out the data is from the mean
- And they use **every single value** to calculate that spread

### Variance Formula
#### Population Variance

$\Large\sigma^2 = \frac{1}{N}\sum\limits_{i=1}^{N}(x_i - \mu)^2$

Where:

- $x_i$ =  each individual value
- $\mu$ =  the population mean
- N     =  the population size

#### Sample Variance

$\Large s^2 = \frac{1}{n - 1}\sum\limits_{i=1}^{n}(x_i - \bar{x})^2$

Where:

- $x_i$ =      each individual value
- $\bar{x}$ =  the sample mean
- n     =      the sample size

### Why not use Raw Deviations?

Because, that would give us a result of zero!

### Step-by-Step Variance Example

For these five values, calculate the variance:<br/>
[4, 5, 7, 10, 14]

First, compute the mean:

$\bar{x} = \frac{4+5+7+10+14}{5} = 8$

Start by building the table:

<table>
    <tr>
        <th>value $\bar{x}$</th>
        <th>mean</th>
        <th>value-mean</th>
        <th>(value-mean)^2</th>
    </tr>
    <tr style="text-align:right;">
        <td>4</td><td>8</td><td>-4</td><td>16</td>
    </tr>
    <tr style="text-align:right;">
        <td>5</td><td>8</td><td>-3</td><td>9</td>
    </tr>
    <tr style="text-align:right;">
        <td>7</td><td>8</td><td>-1</td><td>1</td>
    </tr>
    <tr style="text-align:right;">
        <td>10</td><td>8</td><td>2</td><td>4</td>
    </tr>
    <tr style="text-align:right;">
        <td>14</td><td>8</td><td>6</td><td>36</td>
    </tr>
    <tr style="text-align:right;">
        <td colspan="4">66</td>
    </tr>
</table>

So the population variance is: $\frac{66}{5} = 13.2$

Calculate the population variance using Python:

In [7]:
import numpy as np

data = [4, 5, 7, 10, 14]

# Population Variance
# For sample variance use ddof=1
variance = np.var(data)
sd = np.std(data)
print(f"Variance \t\t: {variance}")
print(f"Standard Deviation \t: {sd:.2f}")

Variance 		: 13.2
Standard Deviation 	: 3.63


### Standard Deviation

Since variance is in squared units, we take the square root of the variance to calculate the **standard deviation**

$\sigma = \sqrt{\sigma^2} = \sqrt{13.2} = 3.63$

This tells us that the **average distance from the mean** is 3.63 units.

### Why Variance and Standard Deviation matter in Machine Learning

<ol style="font-size:18pt;">
    <li>
        Loss functions and error metrics
        <ul>
            <li>Mean Squared Error (MSE)</li>
            <li>Root Mean Squared Error (RMSE)</li>
        </ul>
    </li>
    <li>
        Model stability and noise. A high variance may be due to:
        <ul>
            <li>Data that are very spread out</li>
            <li>Strong noise affecting the performance of the model</li>
        </ul>
    </li>
    <li>
        Feature variability and usefulness
        <ul>
            <li>An almost zero variance, means that this feature doesn't add information to the model</li>
            <li>A very high variance may indicate strong noise</li>
        </ul>
    </li>
    <li>
        Outliers and data quality
        <ul>
            <li>Flag unusual data points</li>
            <li>Investigate or cap extreme values</li>
            <li>Prepare the ground for using z scores</li>
        </ul>
    </li>
</ol>
